core/src/main/scala/kafka/server/FetchSession.scala

```diff
@@ -0,0 +1,645 @@
+      abortedTransactions = respData.abortedTransactions
+    }
+    if (respData.error.code() != 0) {
+      // Partitions with errors are always included in the response.
+      // We also set the cached lastStableOffset to an invalid offset, -1.
+      // This ensures that when the error goes away, we re-send the partition.
+      lastStableOffset = -1
+      mustRespond = true
+    }
+    mustRespond
+  }
+
+  override def hashCode() = Objects.hash(topic, new Integer(partition))
+
+  def canEqual(that: Any) = that.isInstanceOf[CachedPartition]
+
+  override def equals(that: Any): Boolean =
+    that match {
+      case that: CachedPartition => that.canEqual(this) &&
+        this.topic.equals(that.topic) &&
+        this.partition.equals(that.partition)
+      case _ => false
+    }
+}
+
+/**
+  * The fetch session.
+  *
+  * Each fetch session is protected by its own lock, which must be taken before mutable
+  * fields are read or modified.  This includes modification of the session partition map.
+  *
+  * @param id         The unique fetch session ID.
+  * @param privileged True if this session is privileged.  Sessions crated by followers
+  *                   are privileged; sesssion created by consumers are not.
+  * @param partMap    The CachedPartitionMap.
+  * @param creationMs The time in milliseconds when this session was created.
+  * @param lastUsedMs The last used time in milliseconds.  This should only be updated by
+  *                   FetchSessionCache#touch.
+  * @param epoch      The fetch session seequence number.
+  */
+case class FetchSession(val id: Int,
+                        val privileged: Boolean,
+                        val partMap: FetchSession.CACHE_MAP,
+                        val creationMs: Long,
+                        var lastUsedMs: Long,
+                        var epoch: Int) {
+  def size() = synchronized {
+    partMap.size()
+  }
+
+  def lastUsedKey() = synchronized {
+    LastUsedKey(lastUsedMs, id)
+  }
+
+  def evictableKey() = synchronized {
+    EvictableKey(privileged, partMap.size(), id)
+  }
+
+  def metadata(ty: FetchType) = synchronized {
+    new JFetchMetadata(ty, id, epoch)
+  }
+
+  def getFetchOffset(topicPartition: TopicPartition) = synchronized {
+    Option(partMap.find(new CachedPartition(topicPartition))).map(_.fetchOffset)
+  }
+
+  // Update the cached partition data based on the request.
+  def update(fetchData: FetchSession.REQ_MAP, reqMetadata: JFetchMetadata) = synchronized {
+    var numAdded = 0
+    var numUpdated = 0
+    var numRemoved = 0
+    fetchData.entrySet().iterator().asScala.foreach(entry => {
+      val topicPart = entry.getKey
+      val reqData = entry.getValue
+      val newCachedPart = new CachedPartition(topicPart, reqData)
+      val cachedPart = partMap.find(newCachedPart)
+      if (cachedPart == null) {
+        partMap.add(newCachedPart)
+        numAdded = numAdded + 1
+      } else {
+        if (reqData.maxBytes == 0) {
```

Comment:
This seems similar to setting everything to 0 when there is a partition error, right?  It would be clearer if our RPC type system supported a more advanced type system.